### ISSUE TITLE
Mystcraft for1.12.2 disabled MCA Villager GUI's and functions

### ISSUE THREAD

**Issue by WilliamBlaze** [XCompWiz/Mystcraft-Issues#258]
Mystcraft for1.12.2 disabled MCA Villager GUI's and functions; as the name implies.
more specifically. MCA's functionms work when you firts load up a world with mystcraft along side it, but about after 10 seconds (or how ever long it takes for a chunk update to occur in your world) MCA's villager functions and GUI's are suddenly disabled by Mystcraft's villager gui/features.
 while i put in a notice on MCA's github about this issue; i'm posting here as well just on the off chance someone on this end might discover how to correct this issue.

**Comment by XCompWiz** [comment]
Would you please link to that issue here?

**Comment by WilliamBlaze** [comment]
https://github.com/WildBamaBoy/minecraft-comes-alive/issues/1042

**Comment by WilliamBlaze** [comment]
http://prntscr.com/k3o5rv
http://prntscr.com/k3o645

Briefly discussed my own personal insight based on my time spent with playing using MCA and other mods in the past, added in those two screen shots from the discord groups's support chat.
I could be wrong entirely, but these are my personal observations as i understand them.
and hopefully having them in context here might later provide some help or at the very least any misconceptions be later corrected.

**Comment by XCompWiz** [comment]
So the villagers have no interactions in Mystcraft is installed? Only the Mystcraft villager gets the gui? And it's the Mystcraft shop GUI?

**Comment by WilliamBlaze** [comment]
not sure about the usrup thing but i checked the MCA configuration file and there is a feature that says
"  B:"Overwrite original villagers"= false  " by default - when i swithced this value to True, it corrected the problem and restored MCA functions, but at the cost of Mystcraft's own villager shop functions being disabled/overwritten with standard villager trades instead - while MCA does have other interactions not related to villager trades, villager trades in general are still used to raise the friendship levels between your player and the individual villagers, which in turn - higher friendship can leader ot bigger and better trades with the business type villagers such as the butchers, librarians, ect - but this feature itself shines more with villagers such as the archers, warriors, and the local guards ( a custom NPC type added by MCA) when you raise the friendship value with these npcs; eventually you can hire them to help you during your adventuring, and they are more proficient at battling mods then your standard MCA villager may ever hope to be, and the MCA villagers are smarter then and slightly more proficient at saving themselves then the "Testificate" type villagers/aka the original big nosed villagers that went "HNNNN" every 2 seconds."

i mean, there is still the old fashion way of getting things randomly in Mystcraft from randomly generated worlds and hoping to find libraries in them in the current released version right?
so its not like players cant correct the conflict themselves by turning on MCA's overwrite feature in the MCA configs. - Not that i've had to do it before but it makes me glad the feature still exists for 1.12.2's version.

still in all,i'm not sure how ya want to address this, as; like i said; there's a minor fix, but it turns off one of Mystcraft's newest features in the process.
Aside from designing an NPC for mystcraft that minecraft doesn't lop in with the traditional villager/npcs, and then just assigning the mystcraft shop GUI's to function for those specialized mystcraft npcs only, i can't see a way to give players both MCA functions and keep Mystcraft shop GUI functions at the same time. but then again i'm not really much of a coder so there might be a way i'm just unaware of.

**Comment by XCompWiz** [comment]
We'd call that a workaround, not a fix. :)
Can you please clearly give answers to my earlier questions directly? They all were targeted to give me different bits of information.

**Comment by WilliamBlaze** [comment]
So the villagers have no interactions in Mystcraft is installed? Only the Mystcraft villager gets the gui? And it's the Mystcraft shop GUI?

----- Mystcraft GUI interactions with the few librarian based villagers they seem to apply to still function as mystcraft intends them to when MCA's Overwrite is set at false (which it is by default in its configuration files for Minecraft Comes Alive) - but all of MCA's GUI functions are disabled by Mystcraft's GUI shop functions. this means even menu interactions not related to the each villager's trade function is disabled while Mystcraft is installed along side MCA.  -- mystcraft GUI's work, but it breaks every other villager based GUI in the process.  

----- Hopeful that answer the above 3 questions as they all kind of lump hand in hand with each other answer wise.

Are there any errors in the log / console when you right click on a MCA villager?

----- Not a single error pops up in the console when right clicking any villager that falls under MCA's normal functions when Mystcraft is installed, to my understanding; if there's no error, their shouldn't be anything causing conflict... but their's conflict going on that MCA's just not registering as an Error int he console/logs, but its clearly an error other wise MCA's GUI functions should still work as intended/normally, right?
___________________________________________________
Personal note: Maybe the Mystcraft event to usurp the normal right click behavior (bring up the shop gui) is casting without ensuring it's a Minecraft villager? Some other problem with the event?

---- That might be the case? all i know is that whatever the conflict is, enabling Overwrite original villagers"= true in MCA's configurations restores the standard GUI and right click functions provided by MCA to all villagers, at the cost of Mystcraft's Shop GUI being disabled in the process. by any villager that had previously been able to access Mystcraft's shop GUI functions prior to the switch from 
Overwrite original villagers"= false 
to
Overwrite original villagers"= true
in MCA's config file


Does that answer all of the questions? I hope that i provided the information required this time around.

**Comment by XCompWiz** [comment]
I asked them that way to get specific information, and it's hard to parse this way.
Let's try a different method. :) Answer each question directly and with only a few words (preferably).
**What happens when you right click a Mystcraft Archivist?**
**What happens when you right click any other Villager?**

If you find it easier, maybe answer with a screenshot, or think how you would describe the screenshot.

_(The personal note was, like tagged, personally for me. For my memory. :P No worries on answering that.)_

>all of MCA's GUI functions are disabled by Mystcraft's GUI shop functions. this means even menu interactions not related to the each villager's trade function is disabled while Mystcraft is installed along side MCA

This is good new information (and probably needs to be extracted from answering the specifically asked questions). I hadn't realized MCA had other GUIs or that they were also not working.  **Are there other MCA features that don't occur when Mystcraft is installed? Are there any MCA features which do?**

If there's no errors on right click, maybe there are some when the MCA features suddenly stop working?
<hr>
Just some notes to help you improve your bug reporting in general:

When hunting bugs, don't guess at what might be the cause, report exactly what you observe. Like a science project.  The guessing comes later, and, in this case, should usually comes from me. ;)
If you want to guess, put it at the end of your issue report, but I'd recommend keep it short.

> as mystcraft intends them to

This is, for lack of a better term, "negative information."  It actually obscures what you observe behind what you think should be happening, without telling what either is.
Try "Right clicking Mystcraft villagers brings up the shop GUI."  If there is any uncertainty, include a screenshot of the Mystcraft villager and the resultant shop UI. (I'm hoping that's what's happening.)

> if there's no error, their shouldn't be anything causing conflict...

Well, we observe an issue, so there clearly is a conflict. We just need to find where it is. :) While the logs often include hints as to what happened, sometimes they don't include anything at all. Lack of error message could just mean the error didn't get logged, not that there isn't one.

And, while we want to get as much observed info as possible in a bug report, I wouldn't worry about thinking any pass at it should get "all the required information", as we don't always know what that is. :) It usually takes a few passes; clarity and good initial information just cuts down on the number of go-arounds. :P

And I hope you don't feel like I'm giving you a hard time of this. I want to fix the issue, and I'm hoping to use the opportunity to teach you how to do reporting "professionally." ;) It's a useful skill, today, and will benefit me, you, and anyone else you report any bugs to in the future. :)

**Comment by WilliamBlaze** [comment]
**_What happens when you right click a Mystcraft Archivist?_**
I haven't found a villager called "Mystcraft Archivist" myself during gameplay, BUT standard Liberian profession villagers seem to have Mystcraft's Shop GUI when in use with MCA and all configuration files are at their default settings.

**_What happens when you right click any other Villager?_**
Absolutely Nothing. Zero functionality for all villagers but "Librarians". Reference the first question for details.

 **_Are there other MCA features that don't occur when Mystcraft is installed? Are there any MCA features which do?_**

the only features of MCA that still function (without editing its configurations file) While Mystcraft is installed are the following:
- Guards protecting villagers from hostile mobs and hostile players.
 - Villagers forming couples and marrying. then creating a family of their own.
  -Villagers and guards repopulating villages that's original set of villagers had been wiped out by hostile mobs and or by the action of player(s).

All MCA features disabled by the use of Mystcraft (again, without changing configuration settings) are the following:

- basic trades
- play to villager social interaction GUIs
-The ability to marry Villagers to each other  manually at a church.
- the ability for the player to marry a villager and raise their own family
- the ability to gift items to villagers
- the ability to turn zombie villagers back into normal villagers (as MCA changes this from splashing a harming potion and tossing an apple at the infected village to simply Gifting them a gold apple to cure them of the zombie infection)
- the ability to bring back your killed NPC family members (a nice MCA feature btw)
- the ability to hire guards, warriors or archers from a villages to assist you on your adventures. as this function requires interactions with the social GUI provided by MCA to attempt this feature.
- the ability to create and expand your own villages (again, due to this function heavily relying on the social interactions with villagers via MCA's social interaction GUI.)

And these are just the functions i can recall off the top of my head. There could be more functions tied to the Social GUI that have completely slipped my mind. 

Side note: --- Idk if im calling it correctly but its the GUI with all the social interactions performed with villages that is added by the MCA mod, such as hug, flirt, gift items, tell joke, trade, hire, wait here at this spot, ect, So i call it a social GUI for the sake of referencing its observed functions



_Also no worries, i'm a tad blunt with my attempts at reporting bugs when i find them; so learning a new skill is most useful :P_

**Comment by XCompWiz** [comment]
This is excellent! :D Great work, and thanks for the effort put into it.

Can you provide a screenshot of a "Librarian" with the Mystcraft shop GUI? Are all librarians getting the Mystcraft shop GUI?
Note: Mystcraft archivists look like librarians but have gold trim.

**Comment by WilliamBlaze** [comment]
**_Are all librarians getting the Mystcraft shop GUI?_**
they were by default settings, this included MCA villager librarians.
will post the screenshot later when i have more time in my day to do so, i'm glad this has helped! :D

**Comment by WilliamBlaze** [comment]
It seems after updating to the newest Mystcraft version released earlier this month that i can't recreate the original issue anymore. i'm beginning to wonder if the bug was ever at all on Mystcraft's end and not caused by another mod in my personal mod pack,  as i have updated a lot of them recently.
Namely these mods http://prntscr.com/k74kfi and any mods needed to run them. along with these 3 recently http://prntscr.com/k74l13

Have any of the mods listed in the screenshot ever had known issues with Mystcraft in the past?
None of these to my knowledge have ever had reported conflicts with the MCA mod _(that doesn't mean conflicts don't exist, just that no one has ever spoken up about said conflicts if they do exist)_
So i'm curious if the older versions of these mods conflicted Mystcraft to lead to the conditions of the MCA GUI bug.

_(With the exception of the Librarians functioning with the Mystcraft shop GUI, i believe i figured out what happened there. Most likely MCA gave all Mystcraft NPC spawned into the world a MCA villager model. This would explain the Librarians in my village all having access to Mystcraft's shop GUI when all other MCA Villager GU functions stopped. It wouldn't be the first time MCA has given its Better models to replace another mod's villager models if they used a model that was based off vanilla minecraft's Villager models.  MCA Is suppose to do this automatically as one of its feature from my understanding. so if the mystcraft villagers used the vanilla model with a retextured skin, this could explain that part of the phenomenon. and would likely mean the bug was entirely on MCA's end - with Mystcraft just having poor timing in my mod folder line up to be added when said bug started. odd as that is, it wouldn't be the first time a mod was added just as another in world bugged out for a bit. XD)_

That all aside, i don't know if we should close down the issue for now or just make a note of it and put this off to the side in case it returns. But i'll let you be the judge of that.

I'm not even sure now if any of the previous input in this topic helped in any of the recent Mystcraft updates that lead up removing the source of the conflict, or if it was just one of those one off bug minecraft bugs that sometimes happen when attempting to build a mod pack. XD 
(when i finally get the set of mods i like completed, i'll begin going around asking permission to release the pack on curseforge. but until then it'll remain on my PC for personal use in my one player worlds. - and will likely remain as such if some folks say no to their mods inclusion.

Simply because i would like to respect the wishes of the content creators if i can help it'

**Comment by XCompWiz** [comment]
I appreciate the effort and practice of asking for permissions to use mods in a pack. :)

I don't know that I touched anything that should have affected this, but interesting it stopped happening. I'm happy to leave the issue open until some resolution is found  :) Thanks for all the work trying to track it down!
